File: go_tracker/models/group_order.py
from datetime import datetime
from enum import IntEnum
from typing import Optional

from pydantic import confloat
from redis_om import Field, JsonModel, Migrator

from .provider import Provider


def get_timestamp():
    return int(datetime.utcnow().timestamp() * 1000)


class OrderStatus(IntEnum):
    UNPAID = 0
    PARTIALLY_PAID = 1
    FULLY_PAID = 2
    SHIPPED = 3
    DELIVERED = 4


class GroupOrder(JsonModel):
    item: str
    provider: Provider
    order_number: str
    order_date: int = Field(default_factory=get_timestamp)
    downpayment_deadline: Optional[int]
    payment_deadline: int
    status: OrderStatus = Field(index=True, default=OrderStatus.UNPAID.value)
    total_balance: confloat(ge=0) = Field(default=0.0)
    remaining_balance: confloat(ge=0) = Field(default=0.0)


Migrator().run()

File: web/app/src/api/types/groupOrder.ts
import { Provider } from "./provider";
import { Moment } from "moment";

export enum OrderStatus {
  UNPAID,
  PARTIALLY_PAID,
  FULLY_PAID,
  SHIPPED,
  DELIVERED,
}

export const ReverseOrderStatus = [
  {
    label: "UNPAID",
    color: "error.main",
  },
  {
    label: "PARTIALLY_PAID",
    color: "warning.main",
  },
  {
    label: "FULLY_PAID",
    color: "success.main",
  },
  {
    label: "SHIPPED",
    color: "secondary.main",
  },
  {
    label: "DELIVERED",
    color: "primary.main",
  },
];

export interface GroupOrder {
  pk: string;
  item: string;
  provider: Provider;
  order_number: string;
  order_date: number;
  downpayment_deadline: number;
  payment_deadline: number;
  status: OrderStatus;
}

export interface GroupOrderForm {
  pk?: string;
  item: string;
  provider: Provider | null;
  order_number: string;
  order_date: Moment | null;
  downpayment_deadline: Moment | null;
  payment_deadline: Moment | null;
  status: OrderStatus;
}

export interface GroupOrderBody {
  pk?: string;
  item: string;
  provider: Provider | null;
  order_number: string;
  order_date: number;
  downpayment_deadline: number | null;
  payment_deadline: number;
  status: OrderStatus;
}
